sql/logictest: TestParallel failed

sql/logictest.TestParallel [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14567878?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14567878?buildTab=artifacts#/) on release-23.2.3-rc @ [d57701520e29e5226a1e3832b02ecf2810da40bd](https://github.com/cockroachdb/cockroach/commits/d57701520e29e5226a1e3832b02ecf2810da40bd):


```
	NOTICE: .ClusterSettings() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().ClusterSettings() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/sql/logictest/parallel_test.go:178: ((*parallelTest).setup)
	NOTICE: .ClusterSettings() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().ClusterSettings() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/setup with config : 2 tests, 0 failures
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 1 tests, 0 failures
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 0 tests, 0 failures
    parallel_test.go:85: 
        testdata/parallel_test/subquery_retry_multinode/txn:9: 
        expected success, but found
        (40003) result is ambiguous: error=TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=7f6e870c key=/Tenant/10/Table/106/1/954965585192878084/0 iso=Serializable pri=0.02014255 epo=0 ts=1711502768.527534781,1 min=1711502768.388751022,0 seq=3} lock=true stat=PENDING rts=1711502768.516643055,0 wto=false gul=1711502768.888751022,0 [propagate] (last error: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=7f6e870c key=/Tenant/10/Table/106/1/954965585192878084/0 iso=Serializable pri=0.02014255 epo=0 ts=1711502768.527534781,2 min=1711502768.388751022,0 seq=3} lock=true stat=PENDING rts=1711502768.516643055,0 wto=false gul=1711502768.888751022,0)
        dist_sender.go:3050: in sendToReplicas()
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 0 tests, 0 failures
    parallel_test.go:85: 
        testdata/parallel_test/subquery_retry_multinode/txn:9: 
        expected success, but found
        (40003) result is ambiguous: error=TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=9f294017 key=/Tenant/10/Table/106/1/954965585213620227/0 iso=Serializable pri=0.00827606 epo=0 ts=1711502768.527534781,2 min=1711502768.389020036,0 seq=3} lock=true stat=PENDING rts=1711502768.527534781,0 wto=false gul=1711502768.889020036,0 [propagate] (last error: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=9f294017 key=/Tenant/10/Table/106/1/954965585213620227/0 iso=Serializable pri=0.00827606 epo=0 ts=1711502768.527534781,2 min=1711502768.389020036,0 seq=3} lock=true stat=PENDING rts=1711502768.527534781,0 wto=false gul=1711502768.889020036,0)
        dist_sender.go:3050: in sendToReplicas()
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 1 tests, 0 failures
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/txn with config : 0 tests, 0 failures
    parallel_test.go:85: 
        testdata/parallel_test/subquery_retry_multinode/txn:9: 
        expected success, but found
        (40003) result is ambiguous: error=TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=27785b27 key=/Tenant/10/Table/106/1/954965585221419013/0 iso=Serializable pri=0.02867952 epo=0 ts=1711502768.527534781,2 min=1711502768.388961654,0 seq=3} lock=true stat=PENDING rts=1711502768.527132080,0 wto=false gul=1711502768.888961654,0 [propagate] (last error: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=27785b27 key=/Tenant/10/Table/106/1/954965585221419013/0 iso=Serializable pri=0.02867952 epo=0 ts=1711502768.527534781,2 min=1711502768.388961654,0 seq=3} lock=true stat=PENDING rts=1711502768.527132080,0 wto=false gul=1711502768.888961654,0)
        dist_sender.go:3050: in sendToReplicas()
[01:26:08] --- done: testdata/parallel_test/subquery_retry_multinode/final with config : 2 tests, 0 failures
    parallel_test.go:253: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/556c3b72a9585bf329eaad0cafbabdf0/logTestParallel_subquery_retry_multinode1358601955
I240327 01:26:08.850159 12983 sql/logictest/parallel_test.go:258  [-] 2  1 out of 3 parallel tests failed
    --- FAIL: TestParallel/subquery_retry_multinode (33.63s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #120974 sql/logictest: TestParallel failed [C-test-failure O-robot T-sql-queries branch-master release-blocker]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParallel.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37105

--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -1969,6 +1969,11 @@ func slowReplicaRPCWarningStr(
 		dur.Seconds(), attempts, ba, ba.Replica, resp)
 }
 
+// ProxyFailedWithSendError is a marker to indicate a proxy request failed with
+// a sendError. Node.maybeProxyRequest specifically excludes this error from
+// propagation over the wire and instead return the NLHE from local evaluation.
+var ProxyFailedWithSendError = kvpb.NewErrorf("proxy request failed with send error")
+
 // sendPartialBatch sends the supplied batch to the range specified by the
 // routing token.
 //
@@ -2094,15 +2099,47 @@ func (ds *DistSender) sendPartialBatch(
 			tBegin = time.Time{} // prevent reentering branch for this RPC
 		}
 
+		if ba.ProxyRangeInfo != nil {
+			// On a proxy request (when ProxyRangeInfo is set) we always return
+			// the response immediately without retries. Proxy requests are
+			// retried by the remote client, not the proxy node. If the error
+			// contains updated range information from the leaseholder and is
+			// normally retry, the client will update its cache and routing
+			// information and decide how to proceed. If it decides to retry
+			// against this proxy node, the proxy node will see the updated
+			// range information on the retried request, and apply it to its
+			// cache in the SyncTokenAndMaybeUpdateCache call a few lines above.
+			//
+			// TODO(baptist): Update the cache on a RangeKeyMismatchError before
+			// returning the error.
+			if err != nil {
+				log.VEventf(ctx, 2, "failing proxy request after error %s", err)
+				reply = &kvpb.BatchResponse{}
+				if IsSendError(err) {
+					// sendErrors shouldn't[1] escape from the DistSender. If
+					// routing the request resulted in a sendError, we intercept
+					// it here and replace it with a ProxyFailedWithSendError.
+					//
+					// [1] sendErrors are a mechanism for sendToReplicas to
+					// communicate back that the routing information used to
+					// route the request was stale. Therefore, the cache needs
+					// to be flushed and the request should be retried with
+					// fresher information. However, this is specific to the
+					// proxy node's range cache -- not the remote client's range
+					// cache. As such, returning the sendError back to the
+					// remote client is nonsensical.
+					return response{pErr: ProxyFailedWithSendError}
+				} else {
+					reply.Error = kvpb.NewError(kvpb.NewProxyFailedError(err))
+				}
+			}
+			return response{reply: reply}
+		}
+
 		if err != nil {
 			// Set pErr so that, if we don't perform any more retries, the
 			// deduceRetryEarlyExitError() call below the loop includes this error.
 			pErr = kvpb.NewError(err)
-			// Proxy requests are not retried since we not the originator.
-			if ba.ProxyRangeInfo != nil {
-				log.VEventf(ctx, 1, "failing proxy request after error %s", err)
-				break
-			}
 			switch {
 			case IsSendError(err):
 				// We've tried all the replicas without success. Either they're all
@@ -2357,7 +2394,7 @@ func noMoreReplicasErr(ambiguousErr, replicaUnavailableErr, lastAttemptErr error
 	}
 
 	// Authentication and authorization errors should be propagated up rather than
-	// wrapped in a SendError and retried as they are likely to be fatal if they
+	// wrapped in a sendError and retried as they are likely to be fatal if they
 	// are returned from multiple servers.
 	if grpcutil.IsAuthError(lastAttemptErr) {
 		return lastAttemptErr
@@ -2724,14 +2761,14 @@ func (ds *DistSender) sendToReplicas(
 		}
 		if err == nil {
 			if proxyErr, ok := br.Error.GetDetail().(*kvpb.ProxyFailedError); ok {
-				// The server proxy attempt resulted in a non-BatchRequest error, likely
-				// a communication error. Convert the wrapped error into an error on our
-				// side Depending on the type of request and what the error is we may
-				// treat the error an ambiguous error. Clear out the BatchResponse as
-				// the only information it contained was this error.
+				// The server proxy attempt resulted in an error, likely a
+				// communication error. Unwrap the error from the BatchResponse
+				// and set err to the wrapped error. The code below will
+				// correctly determine whether this error is ambiguous based on
+				// the type of error and the type of request.
 				err = proxyErr.Unwrap()
+				log.VEventf(ctx, 2, "proxy error: %s", err)
 				br = nil
-				log.VEventf(ctx, 2, "proxy send error: %s", err)
 			}
 		}
 
@@ -2982,7 +3019,7 @@ func (ds *DistSender) sendToReplicas(
 						// followers will be marked as potential proxy requests and point to
 						// the new leaseholder. We need to try all the replicas again before
 						// giving up.
-						// NB: We reset and retry here because if we release a SendError to
+						// NB: We reset and retry here because if we release a sendError to
 						// the caller, it will call Evict and evict the leaseholder
 						// information we just learned from this error.
 						// TODO(baptist): If sendPartialBatch didn't evict valid range
@@ -3009,7 +3046,7 @@ func (ds *DistSender) sendToReplicas(
 							// regress. As such, advancing through each replica on the
 							// transport until it's exhausted is unlikely to achieve much.
 							//
-							// We bail early by returning a SendError. The expectation is
+							// We bail early by returning a sendError. The expectation is
 							// for the client to retry with a fresher eviction token.
 							log.VEventf(
 								ctx, 2, "transport incompatible with updated routing; bailing early",

--- a/pkg/server/node.go
+++ b/pkg/server/node.go
@@ -1532,11 +1532,17 @@ func (n *Node) maybeProxyRequest(
 	// TODO(baptist): Correctly set up the span / tracing.
 	br, pErr := n.proxySender.Send(ctx, ba)
 	if pErr == nil {
-		log.VEvent(ctx, 2, "proxy succeeded")
+		log.VEvent(ctx, 2, "proxy request completed")
 		return br
+	} else if pErr == kvcoord.ProxyFailedWithSendError {
+		log.VEventf(ctx, 2, "proxy failed with send error %v", pErr)
+		// Use the original error NLHE from local evaluation.
+		return nil
 	}
-	// Wrap the error in a ProxyFailedError. It is unwrapped on the client side
-	// and handled there.
+	// It is rare to get here on a proxy request because wrapping normally
+	// happens in DistSender.sendPartialBatch. Pessimistically wrap the error
+	// and convert this to a ProxyFailedError which may become an ambiguous
+	// error on the other side.
 	log.VEventf(ctx, 2, "proxy attempt resulted in error %v", pErr)
 	br = &kvpb.BatchResponse{}
 	br.Error = kvpb.NewError(kvpb.NewProxyFailedError(pErr.GoError()))

